New SSL Recipe

wanchichen opened the pull request:
## What?

The current SSL recipe/code is very HuBERT-oriented. This PR rewrites it to support any generic SSL task, along with inference with XEUS.

## Major Changes
- Rename previous `ssl1` recipe to `hubert1`. -> HuBERT has many complicated offline steps, so I think its fine to treat it separately. This is also more inline with the actual implementation, which lives in `espnet2/hubert`
- Create new `ssl1` recipe for generic self-supervised learning.
- Create new `ssl` task. The `espnet_model` for this task follows this framework:
    - It contains a `frontend`, `preencoder`, `encoder` as the main modules. 
    - It is trained with 1 or more SSL Losses, which is defined in `espnet2.loss`.
       - Each defined SSL Loss has 2 attributes that need to be defined:
           - `util_attributes`: functions that must be performed for the task to work. For example, masking in hubert and EMA in data2vec. Different tasks may use the same functions, so we need to pre-define this to prevent redundant computation.
           - `required_inputs`: the data needed to perform loss calculation. For example, reconstrunction losses may also require the frontend feats, while HuBERT only needs the encoder output.

## Minor Changes
- Explicitly set `weights_only=False` for `torch.load()`, which is needed in PyTorch 2.6
- Average n best ckpts in deep speed trainer

sw005320 commented:
@simpleoier, I'm not sure you have some bandwidth, but can you review this PR if possible?

sw005320 commented:
@ftshijt, can you review this PR?

sw005320 commented on espnet/nets/pytorch_backend/transformer/embedding.py:
add a unit test

sw005320 replied:
if test/espnet2/ssl/test_ssl_espnet_model.py can handle it, it is fine.

wanchichen replied:
I have a simple test in https://github.com/espnet/espnet/blob/63d38096d088ff11ff6cd290813b1700560bc829/test/espnet2/ssl/test_ssl_espnet_model.py#L26

sw005320 commented on espnet2/train/deepspeed_trainer.py:
@jctian98, can you review the PR related to deepspeed?

sw005320 replied:
@jctian98, this is a reminder.
This PR is very important and I want your review as soon as possible.

sw005320 commented on espnet2/train/collate_fn.py:
This seems to have a lot of changes.
Can you explain these changes?

wanchichen replied:
The most complex part comes from WavLM / Derverb. (lines 92 to 203)

Lines 92 to 128 is just loading the noises 

https://github.com/espnet/espnet/blob/63d38096d088ff11ff6cd290813b1700560bc829/espnet2/train/collate_fn.py#L155 performs reverb augmentation.

https://github.com/espnet/espnet/blob/63d38096d088ff11ff6cd290813b1700560bc829/espnet2/train/collate_fn.py#L175 does WavLM style

And the changes in `__call__()` are from:
- Calling the above functions for augmentation
- Consider cases where we do not have input `text` (ie. any non-HuBERT style SSL training)

sw005320 replied:
I see.
Can you add some comments to them?

sw005320 commented on espnet2/ssl/utils/mask.py:
can you add some simple tests?

sw005320 replied:
Did you add a test?

wanchichen replied:
The original test already covered the usage a bit, but i added another to double check if the masking was done correctly.

sw005320 commented:
The PR looks good to me.
I want to have some clarifications about your changes.
I hope the others also review this PR quickly.

wanchichen commented on egs2/librispeech/ssl1/run.sh:
There is some redundancy, but I thought it would be better to separate them for a better user experience.

HuBERT has additional stages for
- Feature extraction
- K-means
- K-means inference

Which can be skipped for other models. Also, the design of `hubert1` has a loop that runs the following for each Hubert iteration
1. K-means training and inference (above)
2. Collect stats
3. SSL model training

This is good for HuBERT reproducibility since it takes care of all steps from start to finish, but it makes the recipe very complex and hard to modify  (and maybe confusing) for users doing other SSL training techniques

sw005320 replied:
Make sense.
How about adding this description in README.md under the template in ssl/hubert?
Please also add them to the task or other appropriate places in multiple sources to avoid confusions for the users. 
This clarification helps people to understand these two different (but related) tasks.

wanchichen replied:
just added!

simpleoier commented on egs2/TEMPLATE/hubert1/README.md:
```diff
@@ -0,0 +1,3 @@
+# Self-supervised Learning
```
Do you think it's better to add hubert specific information in this file?

simpleoier commented on egs2/librispeech/ssl1/run.sh:
```diff
@@ -7,44 +7,19 @@ set -o pipefail
-./hubert.sh \
-    --ngpu 8 \
+./ssl.sh \
+    --stage 7 \
```
Does this recipe starts from stage 7?

simpleoier reviewed (approved):
Nice! This PR LGTM! Some newly added modules are lack of unit tests. Hope you can add them in the following PRs.

ftshijt commented on egs2/TEMPLATE/ssl1/ssl.sh:
```diff
@@ -0,0 +1,585 @@
+help_message=$(cat << EOF
+Usage: $0 --train-set "<train_set_name>" --valid-set "<valid_set_name>" --test_sets "<test_set_names>"
+
+Options:
```
minor comment: please remember to align the help message to the above arguments.

ftshijt commented on espnet2/ssl/espnet_model.py:
```diff
@@ -0,0 +1,258 @@
+            data["mask_info"] = mask_info
+
+        # Flow Matching
+        # TODO: NOT IMPLEMENTED YET
```
minor comment, leaving TODO with your name ;)

ftshijt reviewed (commented):
Many thanks for the PR!

I left some minor comments, mostly related to documentation side. The design looks great to me.

In addition, it would be super helpful if you can include some examples for using the pre-trained SSL from the recipe for downstream applications. But we may target that in a later PR.

sw005320 commented on espnet2/ssl/espnet_model.py:
```diff
@@ -0,0 +1,235 @@
+
+
+class ESPnetSSLModel(AbsESPnetModel):
+    """A generic SSL model"""
```
Please add more explanations here (or other appropriate spaces).
What kind of SSL models are supported?

wanchichen replied:
Done! For now its only hubert. Data2vec and dinosr are finished implementing, i am just verifying by training now.

jctian98 commented on egs2/librispeech/ssl1/conf/deepspeed.json:
```diff
@@ -0,0 +1,47 @@
+    "train_micro_batch_size_per_gpu": 1,
+      "gradient_accumulation_steps": 1,
+      "gradient_clipping": 5.0,
+      "bf16": {
```
BF16 is only suitable for Ampere and above. Add a comment here?

wanchichen replied:
I put it in the readme instead. I dont think json supports comments natively.

jctian98 commented:
I checked the PR mainly for DeepSpeed-related features. Specifically (1) deepspeed_trainer; (2) model average; (3) related config file. I think the implementation is correct. 

From the reproducibility perspective:
(1) Can we provide guidance on how to install deepspeed and related plugins somewhere? like README.md file. I think people usually get in trouble with plugin compilations. 
(2) We can possibly provide a config file without deepspeed. Just ensure it can run quickly (but remind people we don't exact use that config file)
(3) For the HF models we need some extra setup for activation checkpointing besides in the deepspeed.json. If you also do the checkpointing, could you point out where that setup is? (usually a setup in modeling file)

wanchichen commented:
Thanks everyone for the comments! 

Downstream usage (tokenization and fine-tuning) will be in a separate PR.

> Can we provide guidance on how to install deepspeed and related plugins somewhere? 
I added some details in the librispeech recipe readme. But it will be difficult to make complete documentation since the installation for those packages are very system dependent.

> We can possibly provide a config file without deepspeed.
I added `train_hubert_amp.yaml` as a non-deepspeed config.

>  If you also do the checkpointing, could you point out where that setup is?
Oh i dont do checkpointing here since the models are very small. We can follow up in another PR.

sw005320 commented:
Thanks!!
